In [1]:
# importing data manipulation libraries
import pandas as pd
import numpy as np
# from ydata_profiling import ProfileReport

# importing visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

# configure notebook for inline plotting
%matplotlib inline

# config pandas to display more than 20 columns
pd.set_option('display.max_columns',250)

# set grid style 
sns.set_style('darkgrid')

# Data Understanding
The objective in this section is to learn as much as possible about the our match data. Here, we will get a high-level practical understnading of our data and by the end of the process, we should have a clear undertanding of the structure of the dataset, how to clean the data, the target variables and possible modelling techniques.

In [2]:
# read data into pandas dataframe
data = pd.read_csv('./data/Matches.csv', header=0)

In [3]:
# check the shape of dataframe
data.shape

(6460, 215)

The dataset consists of 6,460 observations and 215 features some of which are our target variables.

In [4]:
# check types of columns
data.dtypes.unique()

array([dtype('int64'), dtype('O'), dtype('float64'), dtype('bool')],
      dtype=object)

The result indicates that our dataset consists of categorical (object), boolean and numeric (float and int) features. Before proceeding further, let's confirm that the dtypes match what is expected in the documentation.

### To do (Create table of columns and their definitions) - Use datawrapper and insert link 



## columns to be used in modelling 
- total goal count (will be target variable for objective two)
- team_a_cards_num	team_b_cards_num(objective three)
- odds_ft_1	odds_ft_x	odds_ft_2
- winningTeam (objective one )- re-code column to 1x2
- home_ppg	away_ppg
- team_a_xg_prematch	team_b_xg_prematch	total_xg_prematch

target columns 
- total goal count
- total cards(obtain by summing team a and b cards)
- winningTeam 1x2

features
- odds_ft_1	odds_ft_x	odds_ft_2
- home_ppg	away_ppg
- team_a_xg_prematch	team_b_xg_prematch	total_xg_prematch


In [5]:
# inspect the first few rows
data.head(20)

id  homeID  awayID     season    status  roundID  game_week  \
0   2155     150     108  2016/2017  complete       19          1   
1   2156     145     154  2016/2017  complete       19          1   
2   2157     143     142  2016/2017  complete       19          1   
3   2158     144      92  2016/2017  complete       19          1   
4   2159     147     141  2016/2017  complete       19          1   
5   2160     146     155  2016/2017  complete       19          1   
6   2161      93     156  2016/2017  complete       19          1   
7   2162     148     149  2016/2017  complete       19          1   
8   2163      59     151  2016/2017  complete       19          1   
9   2164     152     153  2016/2017  complete       19          1   
10  2165     149     146  2016/2017  complete       19          2   
11  2166     141      93  2016/2017  complete       19          2   
12  2167     154     150  2016/2017  complete       19          2   
13  2168      92     143  2016/2017  complete       19          2   
14  2169     155     152  2016/2017  complete       19          2   
15  2170     142     144  2016/2017  complete       19          2   
16  2171     145     151  2016/2017  complete       19          2   
17  2172     108      59  2016/2017  complete       19          2   
18  2173     156     147  2016/2017  complete       19          2   
19  2174     153     148  2016/2017  complete       19          2   

    revised_game_week           homeGoals                   awayGoals  \
0                  -1      ['45+1', '57']                      ['47']   
1                  -1                  []                      ['82']   
2                  -1                  []                      ['74']   
3                  -1               ['5']                      ['59']   
4                  -1              ['11']                      ['67']   
5                  -1              ['58']                       ['9']   
6                  -1         ['4', '87']                      ['71']   
7                  -1              ['69']          ['40', '59', '64']   
8                  -1  ['31', '64', '75']  ['45+1', '49', '56', '63']   
9                  -1        ['47', '89']                      ['77']   
10                 -1        ['36', '52']                          []   
11                 -1              ['49']  ['27', '36', '86', '90+5']   
12                 -1                  []              ['79', '90+2']   
13                 -1              ['82']                          []   
14                 -1              ['55']                ['80', '87']   
15                 -1               ['9']              ['45+2', '60']   
16                 -1         ['2', '37']                          []   
17                 -1                  []                          []   
18                 -1              ['71']                ['13', '45']   
19                 -1              ['85']                          []   

    homeGoalCount  awayGoalCount  totalGoalCount  team_a_corners  \
0               2              1               3               5   
1               0              1               1               7   
2               0              1               1               3   
3               1              1               2               5   
4               1              1               2               9   
5               1              1               2               6   
6               2              1               3              10   
7               1              3               4               4   
8               3              4               7               5   
9               2              1               3               7   
10              2              0               2               4   
11              1              4               5               7   
12              0              2               2               8   
13              1              0               1              10   
14

In [6]:
#columns to be used for modelling 
#goals
#


In [7]:
# inspect the last rows
data.tail()

id  homeID  awayID     season      status  roundID  game_week  \
6455  6689328     143     158  2023/2024  incomplete   100543         38   
6456  6689329     151     223  2023/2024  incomplete   100543         38   
6457  6689330     271     162  2023/2024  incomplete   100543         38   
6458  6689331      93     153  2023/2024  incomplete   100543         38   
6459  6689332     251      92  2023/2024  incomplete   100543         38   

      revised_game_week homeGoals awayGoals  homeGoalCount  awayGoalCount  \
6455                 -1        []        []              0              0   
6456                 -1        []        []              0              0   
6457                 -1        []        []              0              0   
6458                 -1        []        []              0              0   
6459                 -1        []        []              0              0   

      totalGoalCount  team_a_corners  team_b_corners  totalCornerCount  \
6455               0              -1              -1                -1   
6456               0              -1              -1                -1   
6457               0              -1              -1                -1   
6458               0              -1              -1                -1   
6459               0              -1              -1                -1   

      team_a_offsides  team_b_offsides  team_a_yellow_cards  \
6455               -1               -1                   -1   
6456               -1               -1                   -1   
6457               -1               -1                   -1   
6458               -1               -1                   -1   
6459               -1               -1                   -1   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
6455                   -1                -1                -1   
6456                   -1                -1                -1   
6457                   -1                -1                -1   
6458                   -1                -1                -1   
6459                   -1                -1                -1   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
6455                    -1                    -1                     -1   
6456                    -1                    -1                     -1   
6457                    -1                    -1                     -1   
6458                    -1                    -1                     -1   
6459                    -1                    -1                     -1   

      team_b_shotsOffTarget  team_a_shots  team_b_shots  team_a_fouls  \
6455                     -1            -1            -1            -1   
6456                     -1            -1            -1            -1   
6457                     -1            -1            -1            -1   
6458                     -1            -1            -1            -1   
6459                     -1            -1            -1            -1   

      team_b_fouls  team_a_possession  team_b_possession  refereeID  \
6455            -1                 -1                 -1        NaN   
6456            -1                 -1                 -1        NaN   
6457            -1                 -1                 -1        NaN   
6458            -1                 -1                 -1        NaN   
6459            -1                 -1                 -1        NaN   

      coach_a_ID  coach_b_ID                           stadium_name  \
6455        -1.0        -1.0                 Selhurst Park (London)   
6456        -1.0        -1.0                    Anfield (Liverpool)   
6457        -1.0        -1.0  Kenilworth Road (Luton, Bedfordshire)   
6458        -1.0        -1.0            Etihad Stadium (Manchester)   
6459        -1.0        -1.0               Bramall Lane (Sheffield)   

     stadium_location  team_a_cards_num  team_b_cards_num  odds_ft_1  \
6455              NaN                -1                -1   

In [8]:
# inspect random sample of rows
data.sample(5)

id  homeID  awayID     season    status  roundID  game_week  \
1892    49339     217      59  2017/2018  complete      243         38   
3995   753229     149     221  2009/2010  complete    53608         20   
5331  1308277     158     157  2021/2022  complete    72035          2   
1783    49230      93     108  2017/2018  complete      243         27   
6119  6688990     251     144  2023/2024  complete   100543          4   

      revised_game_week                       homeGoals     awayGoals  \
1892                 -1                              []        ['38']   
3995                 -1  ['28', '32', '44', '50', '75']            []   
5331                 -1                  ['45+3', '62']            []   
1783                 -1   ['3', '48', '53', '77', '90']        ['24']   
6119                 -1                  ['33', '45+3']  ['14', '55']   

      homeGoalCount  awayGoalCount  totalGoalCount  team_a_corners  \
1892              0              1               1               7   
3995              5              0               5              -1   
5331              2              0               2               3   
1783              5              1               6               8   
6119              2              2               4               4   

      team_b_corners  totalCornerCount  team_a_offsides  team_b_offsides  \
1892               4                11                0                2   
3995              -1                -1               -1               -1   
5331               4                 7                3                2   
1783               2                10                1                2   
6119               6                10                1                2   

      team_a_yellow_cards  team_b_yellow_cards  team_a_red_cards  \
1892                    1                    0                 0   
3995                    0                    1                 0   
5331                    3                    4                 0   
1783                    2                    3                 0   
6119                    1                    2                 0   

      team_b_red_cards  team_a_shotsOnTarget  team_b_shotsOnTarget  \
1892                 0                     5                     5   
3995                 0                    -1                    -1   
5331                 0                     3                     2   
1783                 0                    12                     3   
6119                 0                     9                     7   

      team_a_shotsOffTarget  team_b_shotsOffTarget  team_a_shots  \
1892                     12                      3            17   
3995                     -1                     -1            -1   
5331                     10                      8            13   
1783                      6                      0            18   
6119                      4                      3            13   

      team_b_shots  team_a_fouls  team_b_fouls  team_a_possession  \
1892             8            12             8                 47   
3995            -1            -1            -1                 -1   
5331            10             8            11                 47   
1783             3            10            12                 70   
6119            10            11            13                 45   

      team_b_possession  refereeID  coach_a_ID  coach_b_ID  \
1892                 53      393.0       403.0       145.0   
3995                 -1      361.0     22031.0       221.0   
5331                 53      721.0       404.0       235.0   
1783                 30      716.0        72.0       204.0   
6119                 55      690.0       406.0       197.0   

                   stadium_name                           stadium_location  \
1892   The John Smith's Stadium  Stadium Way, Huddersfield, West Yorkshire   
3995  Old Trafford (Manchester)                                   

### Columns of interest:
This were are columns that were identified to contain interesting, problematic or missing data and warranted further inspection.

#### Status Column
The first column of interest is the status colum which indicates the whether a match was completed. Looking at the head, tail and random sample snapshots of the df, the status of a match can be either complete or incomplete. Let's inspect unique values in the column:

In [9]:
# check unique values in the status column
data['status'].unique()

array(['complete', 'incomplete', 'suspended'], dtype=object)

Based on the output above, match status can `complete`, `incomplete`, or `suspended`. The dataframe snapshots, particularly the tail, shows that most of the incomplete matches are from the 2023/2024 season. This makes sense because, the 2023/2024 is ongoing and there are bound to be incomplete matches. For our analysis we will want to drop entries where the match status. 

In [10]:
# Check list of incomplete matches
data[data['status'] == 'incomplete' ].sample(10)

id  homeID  awayID     season      status  roundID  game_week  \
6377  6689250     211     143  2023/2024  incomplete   100543         30   
6406  6689279      93     271  2023/2024  incomplete   100543         33   
6400  6689273     148     149  2023/2024  incomplete   100543         33   
6358  6689231     153     145  2023/2024  incomplete   100543         28   
6366  6689239     271     211  2023/2024  incomplete   100543         29   
6327  6689200     211     153  2023/2024  incomplete   100543         25   
6315  6689188      93     144  2023/2024  incomplete   100543         24   
6362  6689235     145     218  2023/2024  incomplete   100543         29   
6452  6689325     209     149  2023/2024  incomplete   100543         38   
6336  6689209     151     271  2023/2024  incomplete   100543         26   

      revised_game_week homeGoals awayGoals  homeGoalCount  awayGoalCount  \
6377                 -1        []        []              0              0   
6406                 -1        []        []              0              0   
6400                 -1        []        []              0              0   
6358                 -1        []        []              0              0   
6366                 -1        []        []              0              0   
6327                 -1        []        []              0              0   
6315                 -1        []        []              0              0   
6362                 -1        []        []              0              0   
6452                 -1        []        []              0              0   
6336                 -1        []        []              0              0   

      totalGoalCount  team_a_corners  team_b_corners  totalCornerCount  \
6377               0              -1              -1                -1   
6406               0              -1              -1                -1   
6400               0              -1              -1                -1   
6358               0              -1              -1                -1   
6366               0              -1              -1                -1   
6327               0              -1              -1                -1   
6315               0              -1              -1                -1   
6362               0              -1              -1                -1   
6452               0              -1              -1                -1   
6336               0              -1              -1                -1   

      team_a_offsides  team_b_offsides  team_a_yellow_cards  \
6377               -1               -1                   -1   
6406               -1               -1                   -1   
6400               -1               -1                   -1   
6358               -1               -1                   -1   
6366               -1               -1                   -1   
6327               -1               -1                   -1   
6315               -1               -1                   -1   
6362               -1               -1                   -1   
6452               -1               -1                   -1   
6336               -1               -1                   -1   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
6377                   -1                -1                -1   
6406                   -1                -1                -1   
6400                   -1                -1                -1   
6358                   -1                -1                -1   
6366                   -1                -1                -1   
6327                   -1                -1                -1   
6315                   -1                -1                -1   
6362                   -1                -1                -1   
6452                   -1                -1                -1   
6336                   -1                -1                -1   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
6377                    -1

In [11]:
# Check seasons from which matches are incomplete
data[data['status'] == 'incomplete' ]['season'].unique()

array(['2023/2024'], dtype=object)

In [12]:
# Check list of incomplete matches
data[data['status'] == 'suspended' ]['season'].unique()

array(['2023/2024'], dtype=object)

The above inspection confirms that all matches with the `incomplete` or `suspended` status are the current season. 

#### game_week and revised_game_week columns
The `revised_game_week` column indicates indicates whether the game-week (the week of the season when the match was played) was revised or not. We observe that for inspected sections of the dataframe, no matches appear to have revised match weeks.

In [13]:
# check for unique values in the 'revised_game_week' column
data['revised_game_week'].unique()

array([-1], dtype=int64)

 -1 is the only unique value in the `revised_game_week` column. Thus this column does not provide any useful information and will be dropped.
 As mentioned before, the `game_week`, indicates the week of the season when the match was played. Typically, the EPL games run from from game-week 1 to game-week 38 (assuming a team plays a single EPL game every week - which is usually the case). 

In [14]:
# check for unique values in the 'game_week' column
data['game_week'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38,  0], dtype=int64)

The above output mostly confirms our that there are typically 38 game weeks in a season. However, there appears to be some entries where the column value is 0.

In [15]:
# check for entries where game_week == 0
data[data['game_week'] == 0]

id  homeID  awayID     season    status  roundID  game_week  \
4180  753648      59     142  2008/2009  complete    53612          0   
4181  753649     153     221  2008/2009  complete    53612          0   
4182  753650     147      92  2008/2009  complete    53612          0   
4183  753651     150     162  2008/2009  complete    53612          0   
4184  753652     144     216  2008/2009  complete    53612          0   
...      ...     ...     ...        ...       ...      ...        ...   
4935  755509     153     158  2007/2008  complete    53621          0   
4936  755510      92     151  2007/2008  complete    53621          0   
4937  755511     156      59  2007/2008  complete    53621          0   
4938  755512     206     216  2007/2008  complete    53621          0   
4939  755513     144     157  2007/2008  complete    53621          0   

      revised_game_week                     homeGoals           awayGoals  \
4180                 -1                         ['4']                  []   
4181                 -1                   ['4', '10']              ['48']   
4182                 -1                  ['71', '87']              ['90']   
4183                 -1                  ['23', '81']               ['8']   
4184                 -1                  ['44', '64']  ['21', '66', '90']   
...                 ...                           ...                 ...   
4935                 -1                   ['8', '88']        ['14', '58']   
4936                 -1                            []        ['69', '74']   
4937                 -1                            []              ['24']   
4938                 -1  ['32', '73', '90+1', '90+3']              ['49']   
4939                 -1            ['28', '70', '82']              ['46']   

      homeGoalCount  awayGoalCount  totalGoalCount  team_a_corners  \
4180              1              0               1              -1   
4181              2              1               3              -1   
4182              2              1               3              -1   
4183              2              1               3              -1   
4184              2              3               5              -1   
...             ...            ...             ...             ...   
4935              2              2               4              -1   
4936              0              2               2              -1   
4937              0              1               1              -1   
4938              4              1               5              -1   
4939              3              1               4              -1   

      team_b_corners  totalCornerCount  team_a_offsides  team_b_offsides  \
4180              -1                -1               -1               -1   
4181              -1                -1               -1               -1   
4182              -1                -1               -1               -1   
4183              -1                -1               -1               -1   
4184              -1                -1               -1               -1   
...              ...               ...              ...              ...   
4935              -1                -1               -1               -1   
4936              -1                -1               -1               -1   
4937              -1                -1               -1               -1   
4938              -1                -1               -1               -1   
4939              -1                -1               -1               -1   

      team_a_yellow_cards  team_b_yellow_cards  team_a_red_cards  \
4180                   -1                   -1                -1   
4181                    2                    1                 0   
4182                    1                    2                 0   
4183                    3                    0                 0   
4184                    2                    2                 0   
...                   ...                  ...

In [16]:
# check seasons where game_week == 0 occur
data[data['game_week'] == 0]['season'].unique()

array(['2008/2009', '2007/2008'], dtype=object)

In [17]:
# check season unique game_week values for the 2008/2009 season
data[data['season'] == '2008/2009']['game_week'].unique()

array([0], dtype=int64)

In [18]:
# check season unique game_week values for the 2007/2008 season
data[data['season'] == '2007/2008']['game_week'].unique()

array([0], dtype=int64)

In [19]:
# check season unique game_week values for the 2009/2010 season
data[data['season'] == '2009/2010']['game_week'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 17, 14, 15, 16,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38], dtype=int64)

Further inspection shows that matches from the `2007/2008` and `2008/2009` seasons have all values in the `game_week` column as 0. `game_week` information seems to have been stardardized starting from `2009/2010` season. If `game_week` data from `2007/2008` and `2008/2009` seasons is to be used during EDA or modelling, it will be important to indicate the correct match-week information.

#### homeGoals and awayGoals columns
This columns consist of arrays that store goal timing data (when goals were scored during the match). Upon closer scrutiny of the dataset, it was observed that the columns `homeGoals` and `homeGoals_timings` contained identical information, as did `awayGoals` and `awayGoals_timings`. In the interest of eliminating redundancy and retaining more descriptive columns, it makes sense to remove the `homeGoals` and `awayGoals` columns.

In [20]:
data[['homeGoals', 'awayGoals', 'homeGoals_timings', 'awayGoals_timings' ]].head()

homeGoals awayGoals homeGoals_timings awayGoals_timings
0  ['45+1', '57']    ['47']    ['45+1', '57']            ['47']
1              []    ['82']                []            ['82']
2              []    ['74']                []            ['74']
3           ['5']    ['59']             ['5']            ['59']
4          ['11']    ['67']            ['11']            ['67']

#### corner & offside columns
Dataframe snapshots (head, tail, sample) showed that some seasons, particularly seasons priror to the `2013/2014` season had missing data (encoded as -1) in the following columns 
- `team_a_corners`
- `team_b_corners`
- `totalCornerCount`
- `team_a_offsides`
- `team_b_offsides`
- `team_a_shotsOnTarget`
- `team_b_shotsOnTarget`
- `team_a_shotsOffTarget`
- `team_b_shotsOffTarget`
- `team_a_shots`
- `team_b_shots`
- `team_a_fouls`
- `team_b_fouls`
- `team_a_possession`
- `team_b_possession`

In [21]:
# group by season and get descriptive stats for columns suspected to contain missing data
missing_data_columns = [
                        'team_a_corners',
                        'team_b_corners',
                        'totalCornerCount',
                        'team_a_offsides',
                        'team_b_offsides',
                        'team_a_shotsOnTarget',
                        'team_b_shotsOnTarget',
                        'team_a_shotsOffTarget',
                        'team_b_shotsOffTarget',
                        'team_a_shots',
                        'team_b_shots',
                        'team_a_fouls',
                        'team_b_fouls',
                        'team_a_possession',
                        'team_b_possession'
                        ]

data.groupby('season')[missing_data_columns].describe()

team_a_corners                                                \
                   count      mean       std  min  25%  50%  75%   max   
season                                                                   
2007/2008          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2008/2009          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2009/2010          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2010/2011          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2011/2012          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2012/2013          380.0  6.252632  3.158833  0.0  4.0  6.0  8.0  15.0   
2013/2014          380.0  6.092105  2.962286  0.0  4.0  6.0  8.0  14.0   
2014/2015          380.0  6.042105  3.243555  0.0  4.0  6.0  8.0  18.0   
2015/2016          380.0  5.931579  3.247983  0.0  4.0  5.0  8.0  17.0   
2016/2017          380.0  5.673684  3.085217  0.0  3.0  5.0  8.0  19.0   
2017/2018          380.0  5.750000  2.999230  0.0  4.0  5.0  8.0  18.0   
2018/2019          380.0  5.726316  3.025632  0.0  4.0  5.0  8.0  17.0   
2019/2020          380.0  5.771053  3.212394  0.0  3.0  5.0  8.0  17.0   
2020/2021          380.0  5.536842  3.026917  0.0  3.0  5.0  7.0  16.0   
2021/2022          380.0  5.584211  3.014611  0.0  3.0  5.0  7.0  15.0   
2022/2023          380.0  5.610526  3.071856  0.0  3.0  5.0  8.0  17.0   
2023/2024          380.0  2.647368  4.138993 -1.0 -1.0  1.0  6.0  17.0   

          team_b_corners                                                 \
                   count      mean       std  min   25%  50%  75%   max   
season                                                                    
2007/2008          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2008/2009          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2009/2010          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2010/2011          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2011/2012          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2012/2013          380.0  4.792105  2.784511  0.0  3.00  4.0  6.0  17.0   
2013/2014          380.0  4.665789  2.695915  0.0  3.00  4.0  6.0  14.0   
2014/2015          380.0  4.663158  2.568782  0.0  3.00  4.0  6.0  13.0   
2015/2016          380.0  4.876316  2.509710  0.0  3.00  5.0  7.0  14.0   
2016/2017          380.0  4.742105  2.705767  0.0  3.00  4.0  6.0  15.0   
2017/2018          380.0  4.544737  2.769148  0.0  2.75  4.0  6.0  13.0   
2018/2019          380.0  4.557895  2.731487  0.0  2.75  4.0  6.0  14.0   
2019/2020          380.0  4.928947  2.839173  0.0  3.00  5.0  7.0  16.0   
2020/2021          380.0  4.613158  2.669151  0.0  3.00  4.0  6.0  13.0   
2021/2022          380.0  4.818421  2.772115  0.0  3.00  5.0  7.0  14.0   
2022/2023          380.0  4.455263  2.814512  0.0  2.00  4.0  6.0  19.0   
2023/2024          380.0  1.873684  3.397165 -1.0 -1.00  1.0  4.0  13.0   

          totalCornerCount                                                    \
                     count       mean       std  min  25%   50%    75%   max   
season                                                                         
2007/2008            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2008/2009            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2009/2010            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2010/2011            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2011/2012            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2012/2013            380.0  11.044737  3.681315  0.0  9.0  11.0  13.25  23.0   
2013/2014            380.0  10.757895  3.522909  2.0  8.0  11.0  13.00  21.0   
2014/2015            380.0  10.705263  3.521135  1.0  8.0  11.0  13.00  23.0   
2015/2016            380.0  10.807895  3.604811  1.0  8.0  10.0  13.00  25.0   
2016/2017            380.0  10.415789  3.442826  2.0  8.0  10.0  13.

The output above confirms that seasons prior to the 2013/2014 season indeed are missing data in the columns identified above. Given that shots, shots on target, and corner kicks are occasions in a football/soccer match when the probability of scoring is high, and therefore probably have significant match outcome prediction power, we will likely retain only data from 2013/2014 season onwards.

#### Referee ID column
The 'refereeID' column is a unique identifier for the referee who officiated a match. Dataframe snapshots revealed that the referee column had some missing values:

In [22]:
data[data['refereeID'].isnull()].shape

(2784, 215)

In [23]:
data[data['refereeID'].isnull()]['season'].unique()

array(['2016/2017', '2015/2016', '2014/2015', '2013/2014', '2017/2018',
       '2012/2013', '2008/2009', '2007/2008', '2023/2024'], dtype=object)

Output indicates that 2,784 entries in the refereeID column have null values. Given that Referees are randomly selected and are often assumed to have very little influence on the match outcome (assuming match officiation is fair), we can safely drop the referee column.

#### Stadium Location
Examination of dataframe shows that the stadium location column has some missing values. since this column has limited relevance, we will add to the list of columns to be dropped.

In [24]:
data[data['stadium_location'].isnull()]['season'].unique()

array(['2016/2017', '2015/2016', '2014/2015', '2013/2014', '2018/2019',
       '2019/2020', '2011/2012', '2010/2011', '2009/2010', '2008/2009',
       '2007/2008', '2020/2021', '2021/2022', '2022/2023', '2023/2024'],
      dtype=object)

#### odds columns
The dataframe snapshots (head, tail, sample) revealed that the values of most of the odds columns, with the exception of `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  columns, were zeros. Odds represent the probability of a given match outcome, therefore, are not expected to be zero.Given that this columns essentially are missing data, we will drop them from the dataset.

In [25]:
# get list of odds columns 
list_of_columns = list(data.columns)

# odds columns
odds_columns = [column for column in list_of_columns if column.startswith('odds')]


In [26]:
# compute the % zero values for each column and store the results in a dictionary
zero_counts = {}
for column in odds_columns:
    zero_counts[column] = (data[column] == 0).sum()/6460

In [27]:
zero_counts

{'odds_ft_1': 0.3238390092879257,
 'odds_ft_x': 0.3238390092879257,
 'odds_ft_2': 0.3238390092879257,
 'odds_ft_over05': 0.6174922600619195,
 'odds_ft_over15': 0.6174922600619195,
 'odds_ft_over25': 0.6164086687306501,
 'odds_ft_over35': 0.6174922600619195,
 'odds_ft_over45': 0.6174922600619195,
 'odds_ft_under05': 0.6174922600619195,
 'odds_ft_under15': 0.6174922600619195,
 'odds_ft_under25': 0.6167182662538699,
 'odds_ft_under35': 0.6173374613003096,
 'odds_ft_under45': 0.6174922600619195,
 'odds_btts_yes': 0.6174922600619195,
 'odds_btts_no': 0.6174922600619195,
 'odds_team_a_cs_yes': 0.6188854489164086,
 'odds_team_a_cs_no': 0.6188854489164086,
 'odds_team_b_cs_yes': 0.6339009287925697,
 'odds_team_b_cs_no': 0.6335913312693499,
 'odds_doublechance_1x': 0.6195046439628483,
 'odds_doublechance_12': 0.6195046439628483,
 'odds_doublechance_x2': 0.6195046439628483,
 'odds_1st_half_result_1': 0.6185758513931888,
 'odds_1st_half_result_x': 0.6187306501547988,
 'odds_1st_half_result_2': 0.

The output above reveals that 32% of entries in the `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  columns have invalid odd data (zeros) while the other odd columns consists of between 62% and 100%  invalid data.Therefore, to ensure model integrity all odds columns except `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  will be dropped from the dataset.Entries with invalid data in the `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  columns will be dropped. 

#### winning_team column
Inspection revealed that this column indicates winning team using the teamid and draws using -1. For EDA it might be useful to create column that indicates home win with `1`, draws with `x` and away wins with 2, which is the industry standard.

#### no_home_away column
We observed tha the `no_home_away` appears to consist of mostly 0 which makes sense as this would only be 1 if a match was played on a neutral ground. This rarely happens in the EPL.

In [28]:
data['no_home_away'].unique()

array([0], dtype=int64)

The above output confirms that no games were played on neutral ground and thus this column can be dropped.
#### Attendance column
In this column we noted that zero attendance or missing attendance was encoded with -1. 

In [29]:
data[data['attendance']==-1]['attendance'].count()

1822

In [30]:
# List of seasons to filter out
season_not_relevant = ['2007/2008', '2008/2009', '2009/2010', '2010/2011', 
                       '2011/2012', '2012/2013', '2023/2024']

# Filter out rows where the "season" column is in the list
filtered_data = data[~data['season'].isin(season_not_relevant)].copy()


# check missing attendance for relevant seasons
filtered_data[filtered_data['attendance']==-1]['attendance'].count()

729

In [31]:
filtered_data[filtered_data['attendance']==-1].groupby('season')['attendance'].count()

season
2013/2014     16
2014/2015     16
2015/2016     16
2016/2017     16
2020/2021     81
2021/2022    263
2022/2023    321
Name: attendance, dtype: int64

Further inspection reveals that for the relevant seasons, seasons to be considered to have sufficient data, very few matches between `2013` and `2019` had missing fan attendance information. While missing attendance data was expected for the `2020/2021` due covid restrictions, missing data in `2021/2022` and `2022/2023` is unusual. 
Since attendance data may be useful in predicting match outcome, missing data could be obtained from past attendance data or stadium capacity can be used with the assumption that the stadium was full.

#### goalTimingDisabled, corner_timings_recorded, & card_timings_recorded columns
This columns consisted mostly of 1's or 0's. API documentation revealed that this columns were used to indicate whether goal timing, corner timing or card timing data was captured. This information has little relevance and can be safely dropped dropped.

#### team_a_fh_corners	team_b_fh_corners	team_a_2h_corners	team_b_2h_corners	corner_fh_count	corner_2h_count	team_a_fh_cards	team_b_fh_cards	team_a_2h_cards	team_b_2h_cards	total_fh_cards	total_2h_cards 

- Review this columns (2016/2017) season showed some missing data for these this season

#### attacks_recorded column
The attacks_recorded column tracks whether or not attack data was captured encoded as -1 (attack data not recorded) and 1 (attack data was recorded). This is a metadata column and thus can be dropped as it does not have any predictive relevance. 

In [32]:
filtered_data['attacks_recorded'].unique()

array([-1,  1], dtype=int64)

In [33]:
filtered_data['attacks_recorded'].value_counts(normalize=True)

 1    0.690263
-1    0.309737
Name: attacks_recorded, dtype: float64

The output above shows that approximately 56% of matches are missing attack data. The affected columns include:
- `team_a_dangerous_attacks`	
- `team_b_dangerous_attacks`	
- `team_a_attacks`	
- `team_b_attacks`

Given that a large fraction of the attack data is missing, and because this data cannot be imputed, dropping the affected columns is a sensible approach. Additionally, given that live match predictions, which rely on attack data,  is not part of the project scope, dropping attack data does not significantly affect the success of the project. 

In [34]:
data['competition_id'].unique()

array([   9,   10,   11,   12,  161,  246, 1625, 2012, 3119, 3121, 3125,
       3131, 3137, 4759, 6135, 7704, 9660], dtype=int64)

ffffffffffffff

In [35]:
# Updating the list of columns to keep, including the additional columns requested
updated_columns_to_keep = [
    'id', 'homeID', 'awayID', 'season', 'homeGoalCount', 'awayGoalCount', 'totalGoalCount',
    'home_ppg', 'away_ppg', 'pre_match_home_ppg', 'pre_match_away_ppg', 
    'pre_match_teamA_overall_ppg', 'pre_match_teamB_overall_ppg', 
    'team_a_cards_num', 'team_b_cards_num', 'odds_ft_1', 'odds_ft_x', 'odds_ft_2',
    'winningTeam', 'team_a_xg_prematch', 'team_b_xg_prematch', 'total_xg_prematch'
]

# Keeping only the updated set of columns in the filtered_data DataFrame
filtered_data = filtered_data[updated_columns_to_keep]

# Creating a new column for the total number of cards
filtered_data['total_cards'] = filtered_data['team_a_cards_num'] + filtered_data['team_b_cards_num']

# Re-coding the 'winningTeam' column to 1x2 format (assuming 1 for home win, 2 for away win, and 0 for draw)
# Modify the logic below as per the actual coding of the 'winningTeam' column in your dataset
filtered_data['winningTeam_1x2'] = filtered_data['winningTeam'].map({1: 1, 2: 2, 'draw': 0})

# Display the first few rows of the modified DataFrame
filtered_data.head()


id  homeID  awayID     season  homeGoalCount  awayGoalCount  \
0  2155     150     108  2016/2017              2              1   
1  2156     145     154  2016/2017              0              1   
2  2157     143     142  2016/2017              0              1   
3  2158     144      92  2016/2017              1              1   
4  2159     147     141  2016/2017              1              1   

   totalGoalCount  home_ppg  away_ppg  pre_match_home_ppg  pre_match_away_ppg  \
0               3      1.47      0.53                 0.0                 0.0   
1               1      1.74      0.74                 0.0                 0.0   
2               1      1.05      0.84                 0.0                 0.0   
3               2      2.26      1.74                 0.0                 0.0   
4               2      0.95      0.89                 0.0                 0.0   

   pre_match_teamA_overall_ppg  pre_match_teamB_overall_ppg  team_a_cards_num  \
0                          0.0                          0.0                 2   
1                          0.0                          0.0                 3   
2                          0.0                          0.0                 2   
3                          0.0                          0.0                 0   
4                          0.0                          0.0                 3   

   team_b_cards_num  odds_ft_1  odds_ft_x  odds_ft_2  winningTeam  \
0                 2       3.41       3.19       2.39          150   
1                 2       2.45       3.22       3.26          154   
2                 2       2.20       3.25       3.80          142   
3                 0       3.13       3.36       2.45           -1   
4                 5       2.49       3.20       3.21           -1   

   team_a_xg_prematch  team_b_xg_prematch  total_xg_prematch  total_cards  \
0                 0.0                 0.0                0.0            4   
1                 0.0                 0.0                0.0            5   
2                 0.0                 0.0                0.0            4   
3                 0.0                 0.0                0.0            0   
4                 0.0                 0.0                0.0            8   

   winningTeam_1x2  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN

In [36]:
print(filtered_data.columns)


Index(['id', 'homeID', 'awayID', 'season', 'homeGoalCount', 'awayGoalCount',
       'totalGoalCount', 'home_ppg', 'away_ppg', 'pre_match_home_ppg',
       'pre_match_away_ppg', 'pre_match_teamA_overall_ppg',
       'pre_match_teamB_overall_ppg', 'team_a_cards_num', 'team_b_cards_num',
       'odds_ft_1', 'odds_ft_x', 'odds_ft_2', 'winningTeam',
       'team_a_xg_prematch', 'team_b_xg_prematch', 'total_xg_prematch',
       'total_cards', 'winningTeam_1x2'],
      dtype='object')


### Feature engineering

In [37]:
# Feature 1: Goal Difference
filtered_data.loc[:, 'goal_difference_per_game'] = abs(filtered_data['homeGoalCount'] - filtered_data['awayGoalCount'])

# Feature 2: Card Difference
filtered_data.loc[:, 'card_difference'] = filtered_data['team_a_cards_num'] - filtered_data['team_b_cards_num']

# Feature 3: Combined Team Strength
filtered_data.loc[:, 'combined_team_strength'] = (filtered_data['home_ppg'] + filtered_data['away_ppg']) / 2


In [38]:
filtered_data.head()

id  homeID  awayID     season  homeGoalCount  awayGoalCount  \
0  2155     150     108  2016/2017              2              1   
1  2156     145     154  2016/2017              0              1   
2  2157     143     142  2016/2017              0              1   
3  2158     144      92  2016/2017              1              1   
4  2159     147     141  2016/2017              1              1   

   totalGoalCount  home_ppg  away_ppg  pre_match_home_ppg  pre_match_away_ppg  \
0               3      1.47      0.53                 0.0                 0.0   
1               1      1.74      0.74                 0.0                 0.0   
2               1      1.05      0.84                 0.0                 0.0   
3               2      2.26      1.74                 0.0                 0.0   
4               2      0.95      0.89                 0.0                 0.0   

   pre_match_teamA_overall_ppg  pre_match_teamB_overall_ppg  team_a_cards_num  \
0                          0.0                          0.0                 2   
1                          0.0                          0.0                 3   
2                          0.0                          0.0                 2   
3                          0.0                          0.0                 0   
4                          0.0                          0.0                 3   

   team_b_cards_num  odds_ft_1  odds_ft_x  odds_ft_2  winningTeam  \
0                 2       3.41       3.19       2.39          150   
1                 2       2.45       3.22       3.26          154   
2                 2       2.20       3.25       3.80          142   
3                 0       3.13       3.36       2.45           -1   
4                 5       2.49       3.20       3.21           -1   

   team_a_xg_prematch  team_b_xg_prematch  total_xg_prematch  total_cards  \
0                 0.0                 0.0                0.0            4   
1                 0.0                 0.0                0.0            5   
2                 0.0                 0.0                0.0            4   
3                 0.0                 0.0                0.0            0   
4                 0.0                 0.0                0.0            8   

   winningTeam_1x2  goal_difference_per_game  card_difference  \
0              NaN                         1                0   
1              NaN                         1                1   
2              NaN                         1                0   
3              NaN                         0                0   
4              NaN                         0               -2   

   combined_team_strength  
0                   1.000  
1                   1.240  
2                   0.945  
3                   2.000  
4                   0.920

In [39]:
# Inspect unique values in the 'winningTeam' column
unique_values_in_winningTeam = filtered_data['winningTeam'].unique()
unique_values_in_winningTeam


array([150, 154, 142,  -1,  93, 149, 151, 152,  92, 144, 145, 147, 153,
       108,  59, 148, 143, 155, 146, 141, 156, 158, 159, 157, 160, 162,
       161, 217, 209, 223, 251, 222, 218, 211], dtype=int64)

## Encode Categorical Variables

In [40]:
# Check data types of columns in filtered_data
data_types = filtered_data.dtypes

# Identify potential categorical columns (usually of type 'object' or 'category')
categorical_columns = data_types[data_types == 'object'].index.tolist()

# Display data types and potential categorical columns
print("Data Types:\n", data_types)

Data Types:
 id                               int64
homeID                           int64
awayID                           int64
season                          object
homeGoalCount                    int64
awayGoalCount                    int64
totalGoalCount                   int64
home_ppg                       float64
away_ppg                       float64
pre_match_home_ppg             float64
pre_match_away_ppg             float64
pre_match_teamA_overall_ppg    float64
pre_match_teamB_overall_ppg    float64
team_a_cards_num                 int64
team_b_cards_num                 int64
odds_ft_1                      float64
odds_ft_x                      float64
odds_ft_2                      float64
winningTeam                      int64
team_a_xg_prematch             float64
team_b_xg_prematch             float64
total_xg_prematch              float64
total_cards                      int64
winningTeam_1x2                float64
goal_difference_per_game         int64
card_differe

Potential Categorical Columns: 'season' as it is an object.

In [46]:
# Assuming 'season' is the only categorical column
# One-hot encoding the 'season' column
encoded_seasons = pd.get_dummies(filtered_data['season'], prefix='season')

# Concatenating the encoded columns back to the original DataFrame and dropping the original 'season' column
filtered_data_encoded = pd.concat([filtered_data.drop('season', axis=1), encoded_seasons], axis=1)

# Display the first few rows to verify
filtered_data_encoded.head()


id  homeID  awayID  homeGoalCount  awayGoalCount  totalGoalCount  \
0  2155     150     108              2              1               3   
1  2156     145     154              0              1               1   
2  2157     143     142              0              1               1   
3  2158     144      92              1              1               2   
4  2159     147     141              1              1               2   

   home_ppg  away_ppg  pre_match_home_ppg  pre_match_away_ppg  \
0      1.47      0.53                 0.0                 0.0   
1      1.74      0.74                 0.0                 0.0   
2      1.05      0.84                 0.0                 0.0   
3      2.26      1.74                 0.0                 0.0   
4      0.95      0.89                 0.0                 0.0   

   pre_match_teamA_overall_ppg  pre_match_teamB_overall_ppg  team_a_cards_num  \
0                          0.0                          0.0                 2   
1                          0.0                          0.0                 3   
2                          0.0                          0.0                 2   
3                          0.0                          0.0                 0   
4                          0.0                          0.0                 3   

   team_b_cards_num  odds_ft_1  odds_ft_x  odds_ft_2  winningTeam  \
0                 2       3.41       3.19       2.39          150   
1                 2       2.45       3.22       3.26          154   
2                 2       2.20       3.25       3.80          142   
3                 0       3.13       3.36       2.45           -1   
4                 5       2.49       3.20       3.21           -1   

   team_a_xg_prematch  team_b_xg_prematch  total_xg_prematch  total_cards  \
0                 0.0                 0.0                0.0            4   
1                 0.0                 0.0                0.0            5   
2                 0.0                 0.0                0.0            4   
3                 0.0                 0.0                0.0            0   
4                 0.0                 0.0                0.0            8   

   winningTeam_1x2  goal_difference_per_game  card_difference  \
0              NaN                         1                0   
1              NaN                         1                1   
2              NaN                         1                0   
3              NaN                         0                0   
4              NaN                         0               -2   

   combined_team_strength  season_2013/2014  season_2014/2015  \
0                   1.000                 0                 0   
1                   1.240                 0                 0   
2                   0.945                 0                 0   
3                   2.000                 0                 0   
4                   0.920                 0                 0   

   season_2015/2016  season_2016/2017  season_2017/2018  season_2018/2019  \
0                 0                 1                 0                 0   
1                 0                 1                 0                 0   
2                 0                 1                 0                 0   
3                 0                 1                 0                 0   
4                 0                 1                 0                 0   

   season_2019/2020  season_2020/2021  season_2021/2022  season_2022/2023  
0                 0                 0                 0                 0  
1                 0                 0                 0                 0  
2                 0                 0                 0                 0  
3                 0                 0                 0                 0  
4                 0                 0                 0                 0

## Scaling 

To ensure that the numerical features in our dataset have the same scale and our models are sensitive to the magnitude of inputs.

Standard Scaling using scikit-learn

In [49]:
from sklearn.preprocessing import StandardScaler

# Selecting numerical columns for scaling (excluding one-hot encoded season columns and identifier columns)
numerical_columns = filtered_data_encoded.select_dtypes(include=['int64', 'float64']).columns
numerical_columns = numerical_columns.drop(['id', 'homeID', 'awayID'])  # Exclude identifier columns

# Applying Standard Scaler
scaler = StandardScaler()
filtered_data_encoded[numerical_columns] = scaler.fit_transform(filtered_data_encoded[numerical_columns])

# Display the first few rows of the scaled DataFrame
filtered_data_encoded.head()


C:\Users\jmwan\anaconda3\lib\site-packages\sklearn\utils\extmath.py:1047: RuntimeWarning: invalid value encountered in divide
  updated_mean = (last_sum + new_sum) / updated_sample_count
C:\Users\jmwan\anaconda3\lib\site-packages\sklearn\utils\extmath.py:1052: RuntimeWarning: invalid value encountered in divide
  T = new_sum / new_sample_count
C:\Users\jmwan\anaconda3\lib\site-packages\sklearn\utils\extmath.py:1072: RuntimeWarning: invalid value encountered in divide
  new_unnormalized_variance -= correction**2 / new_sample_count


id    homeID    awayID  homeGoalCount  awayGoalCount  totalGoalCount  \
0 -0.732438  0.125591 -1.016149       0.359963      -0.181914        0.154549   
1 -0.732437 -0.010330  0.234329      -1.156511      -0.181914       -1.043993   
2 -0.732435 -0.064699 -0.091883      -1.156511      -0.181914       -1.043993   
3 -0.732434 -0.037514 -1.451098      -0.398274      -0.181914       -0.444722   
4 -0.732433  0.044039 -0.119067      -0.398274      -0.181914       -0.444722   

   home_ppg  away_ppg  pre_match_home_ppg  pre_match_away_ppg  \
0 -0.218973 -1.298951           -1.941841           -1.597538   
1  0.299874 -0.882080           -1.941841           -1.597538   
2 -1.026068 -0.683570           -1.941841           -1.597538   
3  1.299136  1.103022           -1.941841           -1.597538   
4 -1.218234 -0.584315           -1.941841           -1.597538   

   pre_match_teamA_overall_ppg  pre_match_teamB_overall_ppg  team_a_cards_num  \
0                    -2.110245                    -2.134071          0.252393   
1                    -2.110245                    -2.134071          1.016611   
2                    -2.110245                    -2.134071          0.252393   
3                    -2.110245                    -2.134071         -1.276043   
4                    -2.110245                    -2.134071          1.016611   

   team_b_cards_num  odds_ft_1  odds_ft_x  odds_ft_2  winningTeam  \
0          0.078050   0.226944  -0.723651  -0.589057     0.651371   
1          0.078050  -0.226025  -0.702015  -0.354013     0.710356   
2          0.078050  -0.343986  -0.680378  -0.208124     0.533401   
3         -1.368725   0.094828  -0.601046  -0.572847    -1.575320   
4          2.248211  -0.207151  -0.716439  -0.367522    -1.575320   

   team_a_xg_prematch  team_b_xg_prematch  total_xg_prematch  total_cards  \
0            -1.28624           -1.291992          -1.330041     0.212528   
1            -1.28624           -1.291992          -1.330041     0.697577   
2            -1.28624           -1.291992          -1.330041     0.212528   
3            -1.28624           -1.291992          -1.330041    -1.727669   
4            -1.28624           -1.291992          -1.330041     2.152725   

   winningTeam_1x2  goal_difference_per_game  card_difference  \
0              NaN                 -0.353164         0.128468   
1              NaN                 -0.353164         0.706196   
2              NaN                 -0.353164         0.128468   
3              NaN                 -1.154849         0.128468   
4              NaN                 -1.154849        -1.026987   

   combined_team_strength  season_2013/2014  season_2014/2015  \
0               -1.100589                 0                 0   
1               -0.412990                 0                 0   
2               -1.258164                 0                 0   
3                1.764409                 0                 0   
4               -1.329789                 0                 0   

   season_2015/2016  season_2016/2017  season_2017/2018  season_2018/2019  \
0                 0                 1                 0                 0   
1                 0                 1                 0                 0   
2                 0                 1                 0                 0   
3                 0                 1                 0                 0   
4                 0                 1                 0                 0   

   season_2019/2020  season_2020/2021  season_2021/2022  season_2022/2023  
0                 0                 0                 0                 0  
1                 0                 0                 0                 0  
2                 0                 0                 0                 0  
3                 0                 0                 0                 0  
4                 0                 0                 0                 0

Min-Max Scaler

In [43]:
# from sklearn.preprocessing import MinMaxScaler

# # Selecting numerical columns for scaling
# # Exclude one-hot encoded season columns and any other non-numerical columns
# numerical_columns = filtered_data_encoded.select_dtypes(include=['int64', 'float64']).columns

# # Applying MinMax Scaler
# scaler = MinMaxScaler()
# filtered_data_encoded[numerical_columns] = scaler.fit_transform(filtered_data_encoded[numerical_columns])

# # Display the first few rows of the scaled DataFrame
# filtered_data_encoded.head()


If you're unsure, you can try both MinMaxScaler and StandardScaler, train your model with each, and evaluate their performance using cross-validation or a separate validation set. This empirical approach will help you determine which scaler works better for your specific dataset and prediction task.

## Splitting the Dataset

In [50]:
from sklearn.model_selection import train_test_split

# Assuming 'totalGoalCount', 'total_cards', and 'winningTeam_1x2' are your target columns
target_columns = ['totalGoalCount', 'total_cards', 'winningTeam_1x2']

# Features (excluding the target columns)
X = filtered_data_encoded.drop(target_columns, axis=1)

# Target variables
y = filtered_data_encoded[target_columns]

# Split the dataset into 80% training and 20% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the shapes of the resulting datasets
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)


Shape of X_train: (3040, 33)
Shape of X_test: (760, 33)
Shape of y_train: (3040, 3)
Shape of y_test: (760, 3)


## Modeling

### Model Selection

1. Linear Regression (Total Goal Count, Total Cards):

Pros: Simple, interpretable, and fast. Can provide insights into linear relationships between features and target variables.

Cons: Assumes a linear relationship, which might not be suitable for complex relationships in the data.


2. Random Forest Regressor (Total Goal Count, Total Cards):

Pros: More flexible than linear regression, can capture non-linear relationships, handles complex datasets well, and less prone to overfitting.

Cons: Less interpretable compared to linear regression.


1. Logistic Regression (Winning Team Prediction):

Pros: Simple, interpretable, and fast. Suitable for binary classification tasks.

Cons: Assumes a linear relationship, may not handle complex relationships as well as non-linear models.


2. Random Forest Classifier (Winning Team Prediction):

Pros: Flexible, handles non-linear relationships, robust to overfitting, and can capture complex decision boundaries.

Cons: More complex and less interpretable compared to logistic regression.

#### Base models


#### Total Goal Count

In [56]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Train a Linear Regression model
lr_model = LinearRegression()
lr_model.fit(X_train, y_train['totalGoalCount'])

# Predict on the test set
y_pred_lr = lr_model.predict(X_test)

# Evaluate the model
mse_lr = mean_squared_error(y_test['totalGoalCount'], y_pred_lr)
r2_lr = r2_score(y_test['totalGoalCount'], y_pred_lr)

# Print metrics with up to 4 significant figures
print("Linear Regression - Mean Squared Error: {:.4g}".format(mse_lr))
print("Linear Regression - R^2 Score: {:.4g}".format(r2_lr))


Linear Regression - Mean Squared Error: 3.329e-30
Linear Regression - R^2 Score: 1


Our model's near-zero MSE and perfect R² score suggest exceptional accuracy but raise overfitting concerns. It's reasonable to re-examine your data and model complexity, and conduct independent tests or cross-validation for validation on unseen data.

#### Total Cards

In [59]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Train a Linear Regression model for Total Cards
lr_model = LinearRegression()
lr_model.fit(X_train, y_train['total_cards'])  # Change target to 'total_cards'

# Predict on the test set
y_pred_lr = lr_model.predict(X_test)

# Evaluate the model for Total Cards
mse_lr = mean_squared_error(y_test['total_cards'], y_pred_lr)  # Change to 'total_cards'
r2_lr = r2_score(y_test['total_cards'], y_pred_lr)  # Change to 'total_cards'

# Print metrics with up to 4 significant figures
print("Linear Regression for Total Cards - Mean Squared Error: {:.4g}".format(mse_lr))
print("Linear Regression for Total Cards - R^2 Score: {:.4g}".format(r2_lr))


Linear Regression for Total Cards - Mean Squared Error: 1.211e-30
Linear Regression for Total Cards - R^2 Score: 1


The Linear Regression model for Total Cards, with an MSE of 1.211e-30 and an R² score of 1, indicates near-perfect predictions but suggests potential overfitting to our training data.

#### Winning Team

In [63]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Train a Logistic Regression model for Winning Team Prediction
logreg_model = LogisticRegression()
logreg_model.fit(X_train, y_train['winningTeam_1x2'])

# Predict on the test set
y_pred_logreg = logreg_model.predict(X_test)

# Evaluate the model
accuracy_logreg = accuracy_score(y_test['winningTeam_1x2'], y_pred_logreg)
report_logreg = classification_report(y_test['winningTeam_1x2'], y_pred_logreg)

# Print evaluation metrics
print("Logistic Regression for Winning Team - Accuracy:", accuracy_logreg)
print("Classification Report:\n", report_logreg)


ValueError: Input y contains NaN.

#### Second Model

#### Total Goal Count

In [58]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Train a Random Forest Regressor
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train['totalGoalCount'])

# Predict on the test set
y_pred_rf = rf_model.predict(X_test)

# Evaluate the model
mse_rf = mean_squared_error(y_test['totalGoalCount'], y_pred_rf)
r2_rf = r2_score(y_test['totalGoalCount'], y_pred_rf)

# Print metrics with up to 4 significant figures
print("Random Forest Regressor - Mean Squared Error: {:.4g}".format(mse_rf))
print("Random Forest Regressor - R^2 Score: {:.4g}".format(r2_rf))


Random Forest Regressor - Mean Squared Error: 0.001853
Random Forest Regressor - R^2 Score: 0.9982


Our Random Forest model shows an MSE of 0.001853 and an R² of 0.9982, indicating high accuracy and excellent fit to the data. The low MSE suggests minimal error in predictions, while the R² close to 1 reflects the model's effectiveness in explaining the variability of the target variable.

#### Total Cards

In [61]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Train a Random Forest Regressor for Total Cards
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train['total_cards'])  # Change target to 'total_cards'

# Predict on the test set
y_pred_rf = rf_model.predict(X_test)

# Evaluate the model for Total Cards
mse_rf = mean_squared_error(y_test['total_cards'], y_pred_rf)  # Change to 'total_cards'
r2_rf = r2_score(y_test['total_cards'], y_pred_rf)  # Change to 'total_cards'

# Print metrics with up to 4 significant figures
print("Random Forest Regressor for Total Cards - Mean Squared Error: {:.4g}".format(mse_rf))
print("Random Forest Regressor for Total Cards - R^2 Score: {:.4g}".format(r2_rf))


Random Forest Regressor for Total Cards - Mean Squared Error: 0.003896
Random Forest Regressor for Total Cards - R^2 Score: 0.9962


The Random Forest model for Total Cards shows an MSE of 0.003896 and an R² of 0.9962, indicating high accuracy and a strong fit to the data.

#### Winning Team

In [64]:
from sklearn.ensemble import RandomForestClassifier

# Train a Random Forest Classifier for Winning Team Prediction
rf_classifier = RandomForestClassifier(random_state=42)
rf_classifier.fit(X_train, y_train['winningTeam_1x2'])

# Predict on the test set
y_pred_rf_class = rf_classifier.predict(X_test)

# Evaluate the model
accuracy_rf_class = accuracy_score(y_test['winningTeam_1x2'], y_pred_rf_class)
report_rf_class = classification_report(y_test['winningTeam_1x2'], y_pred_rf_class)

# Print evaluation metrics
print("Random Forest Classifier for Winning Team - Accuracy:", accuracy_rf_class)
print("Classification Report:\n", report_rf_class)


ValueError: Input y contains NaN.